Adding multi sched support for Scheduler::cycles()

tpai26 opened the pull request:
<!--- Please review your changes in preview mode -->
<!--- Provide a general summary of your changes in the Title above -->

#### Describe Bug or Feature
Currently scheduler.cycles() reads into the default sched_log file to read cycle info. This is to the update the cycle() to read into the sched_log file of the scheduler object which is calling the function.


#### Attach Test and Valgrind Logs/Output
[cycles_selftest.txt](https://github.com/PBSPro/pbspro/files/3455722/cycles_selftest.txt)

[multi_sched_logs.txt](https://github.com/PBSPro/pbspro/files/3448845/multi_sched_logs.txt)
[smoketests_log.txt](https://github.com/PBSPro/pbspro/files/3448846/smoketests_log.txt)

bhroam commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -12176,6 +12176,10 @@ def cycles(self, start=None, end=None, firstN=None, lastN=None):
+        if self.id != "default":
+            day = time.strftime("%Y%m%d", time.localtime())
+            self.logfile = os.path.join(self.attributes['sched_log'], day)
+
```
This might not be the right place to put this.  the PBSService class has init_logfile_path() which sets self.logfile.  Maybe you should overload it and set self.logfile there.  I mean self.logfile is not correctly set for the multischeds.

agrawalravi90 commented on test/tests/selftest/pbs_cycles_test.py:
```diff
@@ -0,0 +1,102 @@
+        self.server.expect(JOB, {'job_state': 'R'}, id=jid1)
+        self.server.expect(JOB, {'job_state': 'R'}, id=jid2)
+
+        cycle = self.scheds['sc'].cycles(start=self.server.ctime, lastN=1)[0]
```
don't you wanna set start to 'st' ?

agrawalravi90 commented on test/tests/selftest/pbs_cycles_test.py:
```diff
@@ -0,0 +1,102 @@
+        self.server.expect(JOB, {'job_state': 'R'}, id=jid1)
+        self.server.expect(JOB, {'job_state': 'R'}, id=jid2)
+
+        cycle = self.scheds['sc'].cycles(start=self.server.ctime, lastN=1)[0]
```
also, don't access the index right-away. Instead, store the value in 'cycles', assert that it's not None and len(cycles) > 0, then do cycle = cycles[0]. That way, the test can fail instead of erroring out if the multisched support doesn't work.

tpai26 replied:
Added a assert to check if the value of cycle is greater than 0 before accessing the index

agrawalravi90 commented on test/tests/selftest/pbs_cycles_test.py:
```diff
@@ -0,0 +1,102 @@
+
+        cycle = self.scheds['sc'].cycles(start=self.server.ctime, lastN=1)[0]
+
+        firstconsidered = cycle.political_order[0]
```
same comment here, don't access the index of political_order directly.

agrawalravi90 commented on test/tests/selftest/pbs_cycles_test.py:
```diff
@@ -0,0 +1,105 @@
+        self.server.expect(JOB, {'job_state': 'R'}, id=jid1)
+        self.server.expect(JOB, {'job_state': 'R'}, id=jid2)
+
+        cycles = self.scheds['sc'].cycles(start=st, lastN=1)
```
don't you want firstN=1 from st time instead of lastN ?

tpai26 replied:
yep, that is right.  I had also forgotten to assert the first considered is jid 1 in the test, added that to the test and attached the new log for it.

tpai26 commented:
I have added the new log for the test I just changed. The other logs are also from the current change to the framework. I did not change the tests, so did not update them again.
